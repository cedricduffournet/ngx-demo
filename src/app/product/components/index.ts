export * from './product-add/product-add.component';
export * from './product-update/product-update.component';
export * from './product-delete/product-delete.component';
export * from './product-form/product-form.component';
export * from './product-item/product-item.component';
export * from './product-items/product-items.component';
export * from './product-details/product-details.component';
