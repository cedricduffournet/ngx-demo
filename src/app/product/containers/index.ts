export * from './product-add-view/product-add-view.component';
export * from './product-update-modal/product-update-modal.component';
export * from './product-delete-modal/product-delete-modal.component';
export * from './product-list-view/product-list-view.component';
export * from './product-selected/product-selected.component';
export * from './product-details-view/product-details-view.component';
