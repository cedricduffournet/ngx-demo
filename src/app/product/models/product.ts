import { schema } from 'normalizr';
import { ProductCategory, productCategorySchema } from '@app/product-category/models/product-category';

export interface Product {
  id: number;
  name: string;
  categories: ProductCategory[];
}

export const productSchema = new schema.Entity('products', {
  categories: [productCategorySchema]
});
