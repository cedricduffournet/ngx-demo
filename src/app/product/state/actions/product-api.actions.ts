import { props, createAction } from '@ngrx/store';

import { NormalizedData } from '@app/shared/models/normalized.model';

export const loadProductSuccess = createAction(
  '[Product API] Load SUCCESS',
  props<{ products: NormalizedData }>()
);

export const loadProductFailure = createAction(
  '[Product API] Load FAILURE',
  props<{ error: any }>()
);

export const addProductSuccess = createAction(
  '[Product API] Add SUCCESS',
  props<{ product: NormalizedData }>()
);

export const addProductFailure = createAction(
  '[Product API] Add FAILURE',
  props<{ error: any }>()
);

export const updateProductSuccess = createAction(
  '[Product API] Update  SUCCESS',
  props<{ product: NormalizedData }>()
);

export const updateProductFailure = createAction(
  '[Product API] Update FAILURE',
  props<{ error: any }>()
);

export const deleteProductSuccess = createAction(
  '[Product API] Delete SUCCESS',
  props<{ id: number }>()
);

export const deleteProductFailure = createAction(
  '[Product API] Delete FAILURE',
  props<{ error: any }>()
);
