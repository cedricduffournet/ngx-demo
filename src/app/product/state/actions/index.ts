import * as ProductUpdateModalActions from './product-update-modal.actions';
import * as ProductAddModalActions from './product-add-modal.actions';
import * as ProductDeleteModalActions from './product-delete-modal.actions';
import * as ProductListViewActions from './product-list-view.actions';
import * as ProductApiActions from './product-api.actions';
import * as ProductActions from './product.actions';
import * as ProductDetailsActions from './product-details.actions';

export {
  ProductListViewActions,
  ProductApiActions,
  ProductUpdateModalActions,
  ProductAddModalActions,
  ProductDeleteModalActions,
  ProductActions,
  ProductDetailsActions
};
