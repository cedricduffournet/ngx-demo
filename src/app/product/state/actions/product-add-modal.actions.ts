import { props, createAction } from '@ngrx/store';

import { Product } from '@app/product/models/product';

export const addProduct = createAction(
  '[Product add  modal] Add',
  props<{ product: Product }>()
);
