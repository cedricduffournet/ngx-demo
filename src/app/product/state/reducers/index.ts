import {
  createSelector,
  createFeatureSelector,
  Action,
  combineReducers
} from '@ngrx/store';

import * as fromRoot from '@app/core/state/reducers';
import * as fromProductCollection from '@app/product/state/reducers/product-collection.reducer';
import * as fromProductEntities from '@app/product/state/reducers/product-entities.reducer';
import * as fromProductCategories from '@app/product-category/state/reducers';

import { Product } from '@app/product/models/product';

export interface ProductsState {
  collection: fromProductCollection.State;
  products: fromProductEntities.State;
}

export interface State extends fromRoot.State {
  products: ProductsState;
}

export function reducers(state: ProductsState | undefined, action: Action) {
  return combineReducers({
    collection: fromProductCollection.reducer,
    products: fromProductEntities.reducer
  })(state, action);
}

export const selectProductsState = createFeatureSelector<ProductsState>(
  'products'
);

export const getProductEntitiesState = createSelector(
  selectProductsState,
  state => state.products
);

export const getProductEntities = createSelector(
  getProductEntitiesState,
  fromProductEntities.getEntities
);

export const getProductEntitiesUpdating = createSelector(
  getProductEntitiesState,
  fromProductEntities.getUpdating
);

export const getProductEntitiesUpdated = createSelector(
  getProductEntitiesState,
  fromProductEntities.getUpdated
);

export const getProductCollectionState = createSelector(
  selectProductsState,
  state => state.collection
);

export const getProductIds = createSelector(
  getProductCollectionState,
  fromProductCollection.getIds
);

export const getProductCollectionAdding = createSelector(
  getProductCollectionState,
  fromProductCollection.getAdding
);

export const getProductCollectionAdded = createSelector(
  getProductCollectionState,
  fromProductCollection.getAdded
);

export const getProductCollectionDeleting = createSelector(
  getProductCollectionState,
  fromProductCollection.getDeleting
);

export const getProductCollectionDeleted = createSelector(
  getProductCollectionState,
  fromProductCollection.getDeleted
);

export const getProductCollectionConfig = createSelector(
  getProductCollectionState,
  fromProductCollection.getConfig
);
export const getProductCollectionTotalItems = createSelector(
  getProductCollectionState,
  fromProductCollection.getTotalItems
);

export const getProducts = createSelector(
  getProductEntities,
  getProductIds,
  (entities, ids) => {
    return ids.map(id => entities[id]);
  }
);

export const getSelectedProductId = createSelector(
  getProductEntitiesState,
  fromProductEntities.getSelectedId
);

export const getSelectedProduct = createSelector(
  getProductEntities,
  getSelectedProductId,
  (productEntities, productId): Product => {
    return productEntities[productId];
  }
);

export const getSelectedProductDenormalized = createSelector(
  getSelectedProduct,
  fromProductCategories.getProductCategoryEntities,
  (selected, productCategories) => {
    return {
      ...selected,
      categories: selected.categories.map(id => productCategories[id])
    };
  }
);
