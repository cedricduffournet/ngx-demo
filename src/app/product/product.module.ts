import { NgModule, ComponentFactoryResolver } from '@angular/core';

import { RouterModule } from '@angular/router';
import { EffectsModule } from '@ngrx/effects';
import { StoreModule } from '@ngrx/store';

import { SharedModule } from '@app/shared/shared.module';
import { PipesModule } from '@app/shared/pipes/pipe.module';

import { ProductCategoryModule } from '@app/product-category/product-category.module';
import { ProductEffects } from '@app/product/state/effects';
import { ProductFacade } from '@app/product/state/product.facade';
import {
  ProductAddComponent,
  ProductUpdateComponent,
  ProductDeleteComponent,
  ProductFormComponent,
  ProductItemComponent,
  ProductItemsComponent,
  ProductDetailsComponent
} from '@app/product/components';
import {
  ProductAddViewComponent,
  ProductUpdateModalComponent,
  ProductDeleteModalComponent,
  ProductListViewComponent,
  ProductSelectedComponent,
  ProductDetailsViewComponent
} from '@app/product/containers';
import { reducers } from '@app/product/state/reducers';
import { ProductGuard } from '@app/product/services/product.guard';
import { CoalescingComponentFactoryResolver } from '@app/coalescing-component-factory-resolver.service';

@NgModule({
  imports: [
    RouterModule,
    SharedModule,
    PipesModule,
    ProductCategoryModule,
    EffectsModule.forFeature([ProductEffects]),
    StoreModule.forFeature('products', reducers)
  ],
  declarations: [
    ProductListViewComponent,
    ProductAddComponent,
    ProductUpdateComponent,
    ProductDeleteComponent,
    ProductFormComponent,
    ProductItemComponent,
    ProductItemsComponent,
    ProductAddViewComponent,
    ProductUpdateModalComponent,
    ProductDeleteModalComponent,
    ProductDetailsComponent,
    ProductSelectedComponent,
    ProductDetailsViewComponent
  ],
  entryComponents: [
    ProductUpdateModalComponent,
    ProductDeleteModalComponent
  ],
  exports: [ProductListViewComponent, ProductAddViewComponent, ProductDetailsViewComponent],
  providers: [ProductFacade, ProductGuard]
})
export class ProductModule {
  // see https://github.com/angular/angular/issues/14324
  constructor(
    coalescingResolver: CoalescingComponentFactoryResolver,
    localResolver: ComponentFactoryResolver
  ) {
    coalescingResolver.registerResolver(localResolver);
  }
}
