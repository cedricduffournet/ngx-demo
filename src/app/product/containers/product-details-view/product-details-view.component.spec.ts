import { Component } from '@angular/core';
import { ComponentFixture, TestBed } from '@angular/core/testing';
import { ActivatedRoute } from '@angular/router';

import { BehaviorSubject } from 'rxjs';
import { provideMockStore, MockStore } from '@ngrx/store/testing';

import { ProductDetailsViewComponent } from '@app/product/containers';
import { ProductDetailsActions } from '@app/product/state/actions';
import * as fromProducts from '@app/product/state/reducers';
import { Store } from '@ngrx/store';

@Component({ selector: 'app-product-selected', template: '' })
class ProductSelectedStubComponent {}

describe('ProductDetailsView', () => {
  let fixture: ComponentFixture<ProductDetailsViewComponent>;
  let component: ProductDetailsViewComponent;
  let activatedRoute: ActivatedRoute;
  let store: MockStore<fromProducts.State>;

  beforeEach(() => {
    TestBed.configureTestingModule({
      declarations: [ProductSelectedStubComponent, ProductDetailsViewComponent],
      providers: [
        provideMockStore(),
        {
          provide: ActivatedRoute,
          useValue: { params: new BehaviorSubject({}) }
        }
      ]
    });

    fixture = TestBed.createComponent(ProductDetailsViewComponent);
    component = fixture.componentInstance;
    activatedRoute = TestBed.get(ActivatedRoute);
    store = TestBed.get(Store);
    jest.spyOn(store, 'dispatch');
  });

  it('should compile', () => {
    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });

  it('should dispatch a ProductDetailsActions.selectProduct action on init', () => {
    fixture.detectChanges();
    const action = ProductDetailsActions.selectProduct({ id: 1 });

    (activatedRoute.params as BehaviorSubject<any>).next({ id: 1 });

    expect(store.dispatch).toHaveBeenLastCalledWith(action);
  });
});
