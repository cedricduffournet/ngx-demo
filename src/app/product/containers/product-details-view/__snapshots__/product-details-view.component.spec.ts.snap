// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductDetailsView should compile 1`] = `
<app-product-details-view
  route={[Function Object]}
  store={[Function MockStore]}
  subscription={[Function Subscriber]}
>
  <app-product-selected />
</app-product-details-view>
`;
